File: interceptor/bsk-hypixel-interceptor.js
/*!
 * BSK Hypixel API 反代拦截器 (JavaScript)
 * ---------------------------------------------------------------
 * 把代码里所有发给 api.hypixel.net 的请求**自动改道**到
 * https://hyp-api.firebounce.today，并把 API Key 换成你自己的 bsk_ Key。
 *
 * 为什么需要它:
 *   很多现成的 Hypixel 工具/网页/脚本里把 base URL 写死在代码深处，
 *   一个个改太麻烦。这个拦截器在**网络层**动手 —— 应用代码一行都不用改。
 *
 * 支持的拦截点:
 *   · fetch()
 *   · XMLHttpRequest
 *
 * 用法（浏览器）:
 *   <script src="bsk-hypixel-interceptor.js"></script>
 *   <script>
 *     BSKHypixel.install({ apiKey: 'bsk_你的Key' });
 *   </script>
 *
 * 用法（Node / Electron）:
 *   const BSKHypixel = require('./bsk-hypixel-interceptor.js');
 *   BSKHypixel.install({ apiKey: process.env.BSK_KEY });
 *
 * ⚠️ Node 的 fetch 是全局的，能拦；但 Node 里如果用了 axios（走 http/https 模块）
 *    拦不到 —— 那种情况请用同目录的 Python 版思路，或者直接把 baseURL 改掉。
 */

(function (root, factory) {
  if (typeof module === 'object' && module.exports) {
    module.exports = factory();
  } else {
    root.BSKHypixel = factory();
  }
})(typeof globalThis !== 'undefined' ? globalThis : this, function () {
  'use strict';

  // ---- 默认配置 ----------------------------------------------------------

  var DEFAULTS = {
    // 反代地址。所有命中目标的请求都会改到这里。
    proxyBase: 'https://hyp-api.firebounce.today',
    // 哪些域名算"原来的 Hypixel API"。
    targetHosts: ['api.hypixel.net'],
    // 你的 bsk_ Key。留空 = 不改 key（需要调用方自己带）。
    apiKey: '',
    // true = **强制**用上面的 Key（把调用方原本带的 key 丢掉）；
    // false = 调用方自己带了 key 就不动。
    forceKey: true,
    // 调试日志。
    debug: false,
    // 拦截到之后回调 (url, 改写后的 url)。想自己记账/上报可以用。
    onIntercept: null,
  };

  var config = Object.assign({}, DEFAULTS);
  var installed = false;
  var originals = { fetch: null, xhrOpen: null, xhrSetHeader: null, xhrSend: null };
  var stats = { intercepted: 0, keyInjected: 0, lastUrl: '', errors: 0 };

  // ---- 纯函数（可单独测试，不依赖 DOM）-----------------------------------

  function log() {
    if (!config.debug) return;
    try {
      var args = Array.prototype.slice.call(arguments);
      args.unshift('[BSK反代]');
      console.log.apply(console, args);
    } catch (e) { /* 忽略 */ }
  }

  /**
   * 这个 URL 是不是"原来的 Hypixel API"？
   * 用 host 精确比较 —— 不能用 indexOf/endswith，否则
   * `api.hypixel.net.evil.com` 也会被当成目标。
   */
  function isTarget(url) {
    var host;
    try {
      host = new URL(url, 'https://placeholder.invalid').hostname.toLowerCase();
    } catch (e) {
      return false;
    }
    return config.targetHosts.some(function (h) {
      return host === String(h).toLowerCase();
    });
  }

  /**
   * 把 URL 从 api.hypixel.net 改写到反代。
   *
   * 只改 **origin**，path / query **原样保留** —— 反代是透明转发，
   * 端点格式跟官方一模一样，所以不需要任何路径映射。
   * 不是目标的 URL 一律原样返回。
   */
  function rewriteUrl(url) {
    if (!url || typeof url !== 'string') return url;
    if (!isTarget(url)) return url;
    try {
      var u = new URL(url, 'https://placeholder.invalid');
      var base = new URL(config.proxyBase);
      u.protocol = base.protocol;
      u.host = base.host;
      return u.toString();
    } catch (e) {
      return url;
    }
  }

  var KEY_HEADERS = ['api-key', 'x-api-key', 'apikey'];

  /** 这个 URL 本来就是发给反代的吗？ */
  function isProxyUrl(url) {
    var host;
    try {
      host = new URL(url, 'https://placeholder.invalid').hostname.toLowerCase();
    } catch (e) {
      return false;
    }
    var base;
    try {
      base = new URL(config.proxyBase).hostname.toLowerCase();
    } catch (e) {
      return false;
    }
    return !!host && host === base;
  }

  /**
   * 这个请求需不需要我们插手（改写 + 补 key）？
   *
   * ★ 两种都要管:
   *   1. 发给**官方** api.hypixel.net 的 -> 改写到反代 + 补 key
   *   2. 本来就发给**反代**的             -> 只补 key（地址不用改）
   *
   * 第 2 种很容易漏: 如果只在"URL 变了"时才补 key, 那
   * "我直接把 baseURL 写成反代, 但想让拦截器自动填 key" 这个
   * 完全合理的用法会**静默失效**（key 不补, 请求 401）。
   */
  function needsHandling(url) {
    return isTarget(url) || isProxyUrl(url);
  }

  function isAuthHeader(name) {
    return String(name || '').toLowerCase() === 'authorization';
  }

  function isKeyHeader(name) {
    return KEY_HEADERS.indexOf(String(name || '').toLowerCase()) !== -1;
  }

  /**
   * 把各种形态的 header 容器统一成 [[名字, 值], ...]。
   * 支持: 普通对象 / Headers / 二维数组。认不出来就返回空数组。
   */
  function headersToPairs(headers) {
    var out = [];
    if (!headers) return out;
    if (typeof Headers !== 'undefined' && headers instanceof Headers) {
      headers.forEach(function (v, k) { out.push([k, v]); });
      return out;
    }
    if (Array.isArray(headers)) {
      headers.forEach(function (p) {
        if (Array.isArray(p) && p.length >= 2) out.push([String(p[0]), String(p[1])]);
      });
      return out;
    }
    if (typeof headers === 'object') {
      Object.keys(headers).forEach(function (k) { out.push([k, headers[k]]); });
    }
    return out;
  }

  /** 调用方是不是**已经**在某个 header 里带了 key（Bearer 也算）。 */
  function hasKeyHeader(headers) {
    return headersToPairs(headers).some(function (p) {
      if (isKeyHeader(p[0])) return String(p[1] || '').trim() !== '';
      if (isAuthHeader(p[0])) return /^bearer\s+\S+/i.test(String(p[1] || ''));
      return false;
    });
  }

  /** URL 的 query 里有没有 ?key=。 */
  function hasKeyInQuery(url) {
    try {
      var u = new URL(url, 'https://placeholder.invalid');
      var k = u.searchParams.get('key');
      return !!(k && k.trim());
    } catch (e) {
      return false;
    }
  }

  /**
   * 决定要不要注入 key、怎么注入。
   *
   * forceKey=true  -> 无条件用我们的 key，并把调用方原来的 key 头**删掉**
   *                   （不删的话有些服务端会优先读旧的那个，等于没换）
   * forceKey=false -> 调用方自己带了就不动
   */
  function planKey(headers, url) {
    var hasKey = hasKeyHeader(headers) || hasKeyInQuery(url);
    if (!config.apiKey) return { inject: false, strip: false, hasKey: hasKey };
    if (hasKey && !config.forceKey) return { inject: false, strip: false, hasKey: hasKey };
    return { inject: true, strip: config.forceKey, hasKey: hasKey };
  }

  /**
   * 对 header 列表执行注入/剥离，返回新的 [[名字, 值], ...]。
   *
   * ★ 输出统一用 `API-Key`（反代三种写法都收），并在浏览器里**换成 X-API-Key**
   *   —— 见 buildFetchHeaders 的说明：跨域请求只有 CORS 白名单里的头才发得出去。
   *
   * strip=true 时会删掉: api-key / x-api-key / apikey 以及 Bearer 的 Authorization。
   */
  function applyKeyToPairs(pairs, plan, headerName) {
    var name = headerName || 'API-Key';
    var out = [];
    pairs.forEach(function (p) {
      if (plan.strip && (isKeyHeader(p[0]) || isAuthHeader(p[0]))) return;
      out.push([p[0], p[1]]);
    });
    if (plan.inject) out.push([name, config.apiKey]);
    return out;
  }

  // ---- fetch -------------------------------------------------------------

  /**
   * 浏览器里跨域请求想带自定义头，那个头必须在服务端
   * Access-Control-Allow-Headers 白名单里，否则预检就过不去。
   * 我们的反代白名单是 `Authorization, X-API-Key, Content-Type` ——
   * **没有 `API-Key`**，所以浏览器场景要用 X-API-Key。
   * Node 没有 CORS，用 API-Key 更贴近官方习惯，无差别。
   */
  function preferredHeaderName() {
    var isBrowser = typeof window !== 'undefined' && typeof window.document !== 'undefined';
    return isBrowser ? 'X-API-Key' : 'API-Key';
  }

  function buildFetchHeaders(original, url) {
    var pairs = headersToPairs(original);
    var plan = planKey(pairs, url);
    if (!plan.inject && !plan.strip) return { headers: original, injected: false, hadKey: plan.hasKey };
    var merged = applyKeyToPairs(pairs, plan, preferredHeaderName());
    // fetch 的 header 参数用 Headers 实例最稳（普通对象无法表示同名多值）。
    var h;
    if (typeof Headers !== 'undefined') {
      h = new Headers();
      merged.forEach(function (p) { h.append(p[0], p[1]); });
    } else {
      h = {};
      merged.forEach(function (p) { h[p[0]] = p[1]; });
    }
    return { headers: h, injected: plan.inject, hadKey: plan.hasKey };
  }

  function installFetch(target) {
    target = target || (typeof globalThis !== 'undefined' ? globalThis : null);
    if (!target || typeof target.fetch !== 'function') return false;
    if (originals.fetch) return true;              // 已经装过
    originals.fetch = target.fetch;

    target.fetch = function (input, init) {
      try {
        var url, request = null;

        if (typeof Request !== 'undefined' && input instanceof Request) {
          request = input;
          url = input.url;
        } else if (input && typeof input === 'object' && typeof input.url === 'string') {
          // 鸭子类型: 别的库造的类 Request 对象
          request = input;
          url = input.url;
        } else {
          url = String(input);
        }

        if (!needsHandling(url)) return originals.fetch.apply(this, arguments);

        var newUrl = rewriteUrl(url);

        if (request && typeof Request !== 'undefined') {
          // ★ 用 Request 当 init 重建，再把 header 覆盖掉 —— 这是规范认可的写法，
          //   能保住 method / body / signal / credentials。
          //   自己手拼 init 很容易把 body 弄丢（尤其流式 body 还要 duplex）。
          var rebuilt = new Request(newUrl, request);
          var pairs = headersToPairs(rebuilt.headers)
            .concat(headersToPairs(init && init.headers));
          var plan = planKey(pairs, newUrl);
          var finalReq = rebuilt;
          if (plan.inject || plan.strip) {
            var h = new Headers();
            applyKeyToPairs(pairs, plan, preferredHeaderName())
              .forEach(function (p) { h.append(p[0], p[1]); });
            finalReq = new Request(rebuilt, { headers: h });
          }
          stats.intercepted++;
          stats.lastUrl = newUrl;
          if (plan.inject) stats.keyInjected++;
          if (config.onIntercept) { try { config.onIntercept(url, newUrl); } catch (e) {} }
          log('fetch (Request)', url, '->', newUrl);
          // init 里剩下的字段（比如 signal）继续生效
          var rest = Object.assign({}, init);
          delete rest.headers;
          return originals.fetch.call(this, finalReq, rest);
        }

        var opts = Object.assign({}, init);
        var r = buildFetchHeaders(opts.headers, newUrl);
        if (r.injected || r.headers !== opts.headers) opts.headers = r.headers;
        if (r.injected) stats.keyInjected++;
        stats.intercepted++;
        stats.lastUrl = newUrl;
        if (config.onIntercept) { try { config.onIntercept(url, newUrl); } catch (e) {} }
        log('fetch', url, '->', newUrl);
        return originals.fetch.call(this, newUrl, opts);
      } catch (e) {
        stats.errors++;
        log('fetch 拦截出错, 原样放行:', e);
        return originals.fetch.apply(this, arguments);
      }
    };
    return true;
  }

  // ---- XMLHttpRequest ----------------------------------------------------

  function installXHR(target) {
    target = target || (typeof globalThis !== 'undefined' ? globalThis : null);
    if (!target || typeof target.XMLHttpRequest !== 'function') return false;
    var XHR = target.XMLHttpRequest;
    if (originals.xhrOpen) return true;

    originals.xhrOpen = XHR.prototype.open;
    originals.xhrSetHeader = XHR.prototype.setRequestHeader;
    originals.xhrSend = XHR.prototype.send;

    XHR.prototype.open = function (method, url) {
      this.__bskOriginalUrl = url;
      this.__bskRewritten = rewriteUrl(String(url));
      if (this.__bskRewritten !== url) {
        stats.intercepted++;
        stats.lastUrl = this.__bskRewritten;
        if (config.onIntercept) { try { config.onIntercept(url, this.__bskRewritten); } catch (e) {} }
        log('xhr', url, '->', this.__bskRewritten);
      }
      var args = Array.prototype.slice.call(arguments);
      args[1] = this.__bskRewritten;
      return originals.xhrOpen.apply(this, args);
    };

    XHR.prototype.setRequestHeader = function (name, value) {
      // 先记下来再决定 —— 因为要等 send 时才知道用户到底带了哪些 key。
      if (!this.__bskHeaders) this.__bskHeaders = [];
      this.__bskHeaders.push([name, value]);
      if (!needsHandling(this.__bskOriginalUrl || '')) {
        return originals.xhrSetHeader.call(this, name, value);
      }
      // 目标是反代: 先什么都不发, 等 send 时统一按策略写。
      var plan = planKey(this.__bskHeaders, this.__bskRewritten || '');
      if (plan.strip && (isKeyHeader(name) || isAuthHeader(name))) return undefined;
      return originals.xhrSetHeader.call(this, name, value);
    };

    XHR.prototype.send = function () {
      if (needsHandling(this.__bskOriginalUrl || '')) {
        var pairs = this.__bskHeaders || [];
        var plan = planKey(pairs, this.__bskRewritten || '');
        if (plan.strip) {
          // setRequestHeader 已经丢掉了旧的 key 头；这里补上我们的。
          if (plan.inject) {
            try {
              originals.xhrSetHeader.call(this, preferredHeaderName(), config.apiKey);
              stats.keyInjected++;
            } catch (e) { stats.errors++; }
          }
        } else if (plan.inject) {
          try {
            originals.xhrSetHeader.call(this, preferredHeaderName(), config.apiKey);
            stats.keyInjected++;
          } catch (e) { stats.errors++; }
        }
      }
      return originals.xhrSend.apply(this, arguments);
    };
    return true;
  }

  // ---- 对外接口 ----------------------------------------------------------

  /**
   * 装上拦截器。
   *
   * options 可选；target 是"要打补丁的那个全局对象"。
   *
   * ★ 第二个参数是给**油猴脚本**用的: 带 @grant 的脚本跑在沙箱里，
   *   直接打 this 的 fetch 是改不到页面本身的；必须把 unsafeWindow
   *   传进来，补丁才落在页面真正用的那个 fetch 上。
   *   普通网页/Node 不用传。
   */
  function install(options, target) {
    if (options) Object.assign(config, options);
    if (!target) target = (typeof globalThis !== 'undefined') ? globalThis : null;
    var a = installFetch(target);
    var b = installXHR(target);
    installed = a || b;
    if (!installed) log('没找到 fetch / XMLHttpRequest, 拦截器没装上');
    return installed;
  }

  function uninstall() {
    var target = (typeof globalThis !== 'undefined') ? globalThis : null;
    if (originals.fetch && target) { target.fetch = originals.fetch; originals.fetch = null; }
    if (originals.xhrOpen && target && target.XMLHttpRequest) {
      target.XMLHttpRequest.prototype.open = originals.xhrOpen;
      target.XMLHttpRequest.prototype.setRequestHeader = originals.xhrSetHeader;
      target.XMLHttpRequest.prototype.send = originals.xhrSend;
      originals.xhrOpen = originals.xhrSetHeader = originals.xhrSend = null;
    }
    installed = false;
  }

  function setConfig(patch) { Object.assign(config, patch || {}); }
  function getConfig() { return Object.assign({}, config); }
  function getStats() { return Object.assign({}, stats); }
  function resetStats() { stats.intercepted = 0; stats.keyInjected = 0; stats.errors = 0; }
  function isInstalled() { return installed; }
  function maskKey(k) {
    k = String(k || '');
    return k.length <= 12 ? k : k.slice(0, 8) + '…' + k.slice(-4);
  }

  return {
    install: install,
    uninstall: uninstall,
    setConfig: setConfig,
    getConfig: getConfig,
    getStats: getStats,
    resetStats: resetStats,
    isInstalled: isInstalled,
    maskKey: maskKey,
    // 纯函数也导出去，方便测试和二次封装
    rewriteUrl: rewriteUrl,
    isTarget: isTarget,
    isProxyUrl: isProxyUrl,
    needsHandling: needsHandling,
    headersToPairs: headersToPairs,
    applyKeyToPairs: applyKeyToPairs,
    planKey: planKey,
    buildFetchHeaders: buildFetchHeaders,
    DEFAULTS: DEFAULTS,
  };
});

File: interceptor/bsk_hypixel_interceptor.py
# -*- coding: utf-8 -*-
"""
BSK Hypixel API 反代拦截器 (Python)
================================================================

把代码里所有发给 ``api.hypixel.net`` 的请求**自动改道**到
``https://hyp-api.firebounce.today``，并把 API Key 换成你自己的 ``bsk_`` Key。

为什么需要它
------------
很多现成的 Hypixel 工具/脚本把 base URL 写死在代码深处，一个个改太麻烦。
这个拦截器在**网络层**动手 —— 业务代码一行都不用改。

支持两种调用方式（装了哪个拦哪个）::

    # requests（绝大多数脚本都用它）
    import bsk_hypixel_interceptor as bsk
    bsk.install(api_key="bsk_你的Key")
    import requests
    requests.get("https://api.hypixel.net/v2/player", params={"uuid": "..."})
    #  ↑ 上面这行会真的打到反代, 并带上你的 key

    # 标准库 urllib
    bsk.install(api_key="bsk_你的Key")
    import urllib.request
    urllib.request.urlopen("https://api.hypixel.net/v2/status?uuid=...")

也可以当上下文管理器用（用完自动还原）::

    with bsk.intercept(api_key="bsk_..."):
        requests.get("https://api.hypixel.net/v2/player", params={"uuid": "..."})

⚠️ 这个模块**只改 URL 和 header**，不碰响应体 —— 反代是透明转发，
   返回格式跟 Hypixel 官方一模一样。
"""

from __future__ import annotations

import threading

__all__ = [
    "install", "uninstall", "intercept", "configure",
    "rewrite_url", "is_target", "is_proxy_url", "needs_key", "apply_key",
    "stats", "reset_stats", "mask_key",
    "DEFAULTS", "PROXY_BASE",
]

PROXY_BASE = "https://hyp-api.firebounce.today"

DEFAULTS = {
    # 反代地址。所有命中目标的请求都会改到这里。
    "proxy_base": PROXY_BASE,
    # 哪些域名算"原来的 Hypixel API"。
    "target_hosts": ("api.hypixel.net",),
    # 你的 bsk_ Key。留空 = 不改 key（需要调用方自己带）。
    "api_key": "",
    # True = **强制**用上面的 Key（把调用方原本带的 key 丢掉）；
    # False = 调用方自己带了 key 就不动。
    "force_key": True,
    # 打印每一步改写（排查问题时开）。
    "debug": False,
}

_cfg = dict(DEFAULTS)
_lock = threading.RLock()
_installed = {"requests": False, "urllib": False}
_originals = {}
_stats = {"intercepted": 0, "key_injected": 0, "errors": 0, "last_url": ""}

# 认这些 header 名（大小写不敏感）当成"调用方带了 key"
_KEY_HEADERS = ("api-key", "x-api-key", "apikey")


# ---------------------------------------------------------------------------
# 纯函数（不依赖第三方库，可单独测试）
# ---------------------------------------------------------------------------

def _log(*a):
    if _cfg.get("debug"):
        print("[BSK反代]", *a)


def _host_of(url):
    """取 URL 的 host（小写，不含端口）。解析不出来返回空串。"""
    try:
        from urllib.parse import urlparse
        return (urlparse(str(url)).hostname or "").lower()
    except Exception:
        return ""


def is_target(url):
    """
    这个 URL 是不是"原来的 Hypixel API"？

    ★ 用 host **精确比较** —— 不能用 in / endswith，否则
      ``api.hypixel.net.evil.com`` 也会被当成目标，等于把你的
      bsk_ Key 主动送给钓鱼站。
    """
    host = _host_of(url)
    if not host:
        return False
    return host in tuple(h.lower() for h in _cfg["target_hosts"])


def rewrite_url(url):
    """
    把 URL 从 api.hypixel.net 改写到反代。

    只改 **host**，path / query **原样保留** —— 反代是透明转发，
    端点格式跟官方一模一样，不需要任何路径映射。
    不是目标的 URL 一律原样返回。
    """
    if not is_target(url):
        return url
    try:
        from urllib.parse import urlparse, urlunparse
        u = urlparse(str(url))
        p = urlparse(_cfg["proxy_base"])
        return urlunparse((p.scheme, p.netloc, u.path, u.params, u.query, u.fragment))
    except Exception:
        return url


def is_proxy_url(url):
    """这个 URL 本来就是发给反代的吗？"""
    h = _host_of(url)
    return bool(h) and h == _host_of(_cfg["proxy_base"])


def needs_key(url):
    """
    这个请求需不需要我们插手（改写 + 补 key）？

    ★ 两种都要管，不能只管第一种:
      1. 发给**官方** api.hypixel.net 的  -> 改写到反代 + 补 key
      2. 本来就发给**反代**的              -> 只补 key（地址不用改）

    第 2 种很容易被漏掉 —— 如果只在"URL 变了"时才补 key，那么
    "我把 baseURL 直接写成反代，但想让拦截器自动填 key" 这个
    完全合理的用法就会静默失效（key 不补，请求 401）。
    """
    return is_target(url) or is_proxy_url(url)


def _has_key(headers):
    """调用方是不是已经带了 key（含 ``Authorization: Bearer``）。"""
    if not headers:
        return False
    try:
        items = headers.items()
    except AttributeError:
        try:
            items = headers
        except Exception:
            return False
    for k, v in items:
        kl = str(k).lower()
        sv = str(v or "")
        if kl in _KEY_HEADERS and sv.strip():
            return True
        if kl == "authorization" and sv.strip().lower().startswith("bearer "):
            return True
    return False


def _has_key_in_query(url):
    """URL 的 query 里有没有 ``?key=``。"""
    try:
        from urllib.parse import urlparse, parse_qs
        q = parse_qs(urlparse(str(url)).query)
        return bool((q.get("key") or [""])[0].strip())
    except Exception:
        return False


def apply_key(headers, url=""):
    """
    按策略给 header 容器加上 / 覆盖上我们的 key，返回**新的 dict**。

    ``force_key=True``  -> 无条件用我们的 key，并把调用方原来的
                           key 头**删掉**（不删的话有些服务端会优先读旧
                           的那个，等于没换）
    ``force_key=False`` -> 调用方自己带了就不动

    headers 可以是 dict / requests 的 CaseInsensitiveDict / 二维可迭代。
    """
    out = {}
    if headers:
        try:
            items = list(headers.items())
        except AttributeError:
            items = list(headers)
        except Exception:
            items = []
        for k, v in items:
            out[str(k)] = v

    key = _cfg.get("api_key") or ""
    if not key:
        return out

    has_key = _has_key(out) or _has_key_in_query(url)
    if has_key and not _cfg.get("force_key"):
        return out

    if _cfg.get("force_key"):
        # ★ 删掉旧的 —— 不能只是"再加一个"，否则两个 key 同时存在时
        #   服务端读哪个是不确定的。
        for name in list(out.keys()):
            nl = name.lower()
            if nl in _KEY_HEADERS:
                del out[name]
            elif nl == "authorization" and str(out[name] or "").strip().lower().startswith("bearer "):
                del out[name]
    out["API-Key"] = key
    _stats["key_injected"] += 1
    return out


# ---------------------------------------------------------------------------
# 装 / 卸
# ---------------------------------------------------------------------------

def _install_requests():
    try:
        import requests
    except ImportError:
        return False
    if _installed["requests"]:
        return True

    orig = requests.Session.request
    _originals["requests"] = orig

    def patched(self, method, url, **kwargs):
        try:
            if needs_key(url):
                new_url = rewrite_url(url)
                kwargs["headers"] = apply_key(kwargs.get("headers"), new_url)
                _stats["intercepted"] += 1
                _stats["last_url"] = new_url
                _log("requests", (method or "GET"), url, "->", new_url)
                url = new_url
        except Exception:
            # 拦截器自己出错**绝不能**影响业务请求 —— 原样放行。
            _stats["errors"] += 1
        return orig(self, method, url, **kwargs)

    requests.Session.request = patched
    _installed["requests"] = True
    return True


def _install_urllib():
    try:
        import urllib.request as ur
        import urllib.error  # noqa: F401  (确保子模块加载)
    except ImportError:
        return False
    if _installed["urllib"]:
        return True

    orig = ur.urlopen
    _originals["urllib"] = orig

    def patched(url, data=None, timeout=None, **kw):
        try:
            if isinstance(url, ur.Request):
                new = rewrite_url(url.full_url)
                if needs_key(url.full_url):
                    headers = dict(url.headers or {})
                    headers = apply_key(headers, new)
                    req = ur.Request(new, data=url.data, headers=headers)
                    for attr in ("origin_req_host", "unverifiable", "method"):
                        try:
                            setattr(req, attr, getattr(url, attr))
                        except Exception:
                            pass
                    _stats["intercepted"] += 1
                    _stats["last_url"] = new
                    _log("urllib(Request)", url.full_url, "->", new)
                    return orig(req, data=None, timeout=timeout, **kw)
            else:
                if needs_key(url):
                    new = rewrite_url(url)
                    # ★ 传**字符串**给 urlopen 是没法加 header 的 —— 必须包成
                    #   Request，否则 key 根本送不出去（请求会 401）。
                    #   urlopen(s) 本来就等价于 urlopen(Request(s))，包一层不改变语义。
                    req = ur.Request(new, data=data, headers=apply_key({}, new))
                    _stats["intercepted"] += 1
                    _stats["last_url"] = new
                    _log("urllib", url, "->", new)
                    return orig(req, data=None, timeout=timeout, **kw)
        except Exception:
            _stats["errors"] += 1
        if timeout is None:
            return orig(url, data=data, **kw)
        return orig(url, data=data, timeout=timeout, **kw)

    ur.urlopen = patched
    _installed["urllib"] = True
    return True


def install(**options):
    """
    装上拦截器。返回装成功的名字列表，比如 ``['requests']``。

    ``install(api_key="bsk_...")``        —— 最常用
    ``install(proxy_base=..., force_key=False)``  —— 自定义
    """
    configure(**options)
    got = []
    if _install_requests():
        got.append("requests")
    if _install_urllib():
        got.append("urllib")
    if not got:
        _log("requests 和 urllib 都没拦到（requests 没装？）")
    else:
        _log("已拦截:", ", ".join(got))
    return got


def uninstall():
    """还原所有补丁。"""
    with _lock:
        if _installed.get("requests") and "requests" in _originals:
            try:
                import requests
                requests.Session.request = _originals["requests"]
            except Exception:
                pass
            _installed["requests"] = False
        if _installed.get("urllib") and "urllib" in _originals:
            try:
                import urllib.request as ur
                ur.urlopen = _originals["urllib"]
            except Exception:
                pass
            _installed["urllib"] = False


class intercept(object):
    """
    当上下文管理器用::

        with bsk.intercept(api_key="bsk_..."):
            ...   # 这里面的请求会走反代
        # 出了 with 自动还原成原来的状态
    """

    def __init__(self, **options):
        self.options = options
        self._before = None

    def __enter__(self):
        self._before = {k: v for k, v in _cfg.items()}
        install(**self.options)
        return self

    def __exit__(self, *exc):
        uninstall()
        with _lock:
            _cfg.update(self._before or {})
        return False


def configure(**options):
    """改配置（不用重装）。"""
    with _lock:
        for k, v in (options or {}).items():
            if k in _cfg:
                _cfg[k] = v
            else:
                raise KeyError("未知配置项: %r（可用: %s）"
                               % (k, ", ".join(sorted(_cfg))))
    return dict(_cfg)


def stats():
    """拦截统计，方便排查"到底拦到没有"。"""
    with _lock:
        return dict(_stats, installed=dict(_installed), config=dict(_cfg))


def reset_stats():
    for k in ("intercepted", "key_injected", "errors"):
        _stats[k] = 0
    _stats["last_url"] = ""


def mask_key(k=None):
    """打日志用的掩码。"""
    k = str(k if k is not None else _cfg.get("api_key") or "")
    return k if len(k) <= 12 else "%s…%s" % (k[:8], k[-4:])
